feat: Allow all config formats in all config locations

danrneal opened the pull request:
## Steps

- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [ ] If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description


## Type of Changes
<!-- Leave the corresponding lines for the applicable type of change: -->
|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |
| ✓  | :hammer: Refactoring  |
| ✓  | :scroll: Docs |

This PR would allow a `pylintrc`, `.pylintrc`, `pyproject.toml`, or `setup.cfg` in any location that config files are stored.  For example, currently `pyproject.toml` will only be read from the project root but not from the home directory.  This PR would allow it to be read from the home directory.  This is helpful when creating a global default config as other projects (such as isort) allow this config format in the home director.  Let me know if this is a welcome change with or without revision or an unwelcome one.

Pierre-Sassoulas commented on pylint/config/find_default_config_files.py:
```diff
@@ -24,40 +24,45 @@ def _cfg_has_config(path):
-    """Find all possible config files."""
-    rc_names = ("pylintrc", ".pylintrc")
-    config_names = rc_names + ("pyproject.toml", "setup.cfg")
+def _get_config_paths(curdir):
```
Could you add typing on the modified functions, please ? :) 
```suggestion
def _get_config_paths(curdir: Union[Path, str]) -> List[Path]:
```

Pierre-Sassoulas reviewed (changes requested):
I think this is a welcome change, and a very important one, but also a very sensible one that could make a lot of CI/run fails if not done correctly. Could you add tests for this, please ? In particular for the get_config_path function.  I'm also going to wait for another review before merging it for this reason.

danrneal commented:
I rebased onto master, added typing, and added a test case.  Thanks for the input.  Let me know if I need more tests or can improve in another way.

Pierre-Sassoulas commented on .gitignore:
```diff
@@ -21,7 +21,9 @@ debian/pylint
 .coverage
 .coverage.*
 .idea
+.vscode
```
These should go into your `.global-gitignore` instead since they are not related to the project itself but rather to your dev environment.

danrneal replied:
I would argue that it is project specific.  Many projects I have worked on, the .vscode/settings.json file is committed to share settings between devs.  In github's semi-official .gitignore files they recommend committing this file.  It is also consistent with the current .gitignore since .idea is in there as well.  So I don't want to put it in my .global-gitignore as it would break other projects, but I can remove it from this .gitignore if still needed.

Pierre-Sassoulas replied:
Fair enough, evidently we did not enforce this rule all the time in the past. Could prevent someone committing it by mistake too.

Pierre-Sassoulas commented on pylint/config/find_default_config_files.py:
```diff
@@ -24,40 +26,45 @@ def _cfg_has_config(path):
+    return paths
+
+
+def find_default_config_files():
```
```suggestion
def find_default_config_files() -> str:
```

danrneal replied:
I'm a newbie at type annotations, thanks for catch. :+1:

Pierre-Sassoulas reviewed (changes requested):
Thank you for taking my review into account. I don't have enough time right now to check that it's not breaking the current expected order of configuration file priority, Please don't hesitate to ping me if I still did not review it properly by october.

danrneal commented:
The only place it may break current order of priority is in the home folder.  Previously it only looked for `.pylintrc`.  Now it looks for `pylintrc` before `.pylintrc`.  In fact it prioritizes configs in the same order in every location now. `pylintrc`, `.pylintrc`, `pyproject.toml`, and lastly for `setup.cfg`.  I updated documentation with the PR to be more clear about priority as well.

Pierre-Sassoulas reviewed (approved):
Sorry for the long time between first fast review and actual review. I don't see a problem here but I'm going to wait for another review before putting it in 2.7.0.

danrneal commented:
Sounds good, just let me know of any feedback and/or changes needed :)

hippo91 commented on doc/user_guide/run.rst:
```diff
@@ -85,26 +85,27 @@ Specifying all the options suitable for your setup and coding
-   providing it has at least one ``tool.pylint.`` section.
-#. ``setup.cfg`` in the current working directory,
-   providing it has at least one ``pylint.`` section
+configuration file valid configuration file (defined below) in the following
```
It seems `configuration file` is redundant.

hippo91 commented on doc/user_guide/run.rst:
```diff
@@ -85,26 +85,27 @@ Specifying all the options suitable for your setup and coding
+#. In the current working directory
 #. If the current working directory is in a Python module, Pylint searches \
-   up the hierarchy of Python modules until it finds a ``pylintrc`` file. \
-   This allows you to specify coding standards on a module-by-module \
+   up the hierarchy of Python modules until it finds a valid configuration \
+   file.  This allows you to specify coding standards on a module-by-module \
    basis.  Of course, a directory is judged to be a Python module if it \
    contains an ``__init__.py`` file.
```
Shouldn't be package instead of module?

hippo91 commented on doc/user_guide/run.rst:
```diff
@@ -85,26 +85,27 @@ Specifying all the options suitable for your setup and coding
+A valid configuration file is one of the following:
 
-#. ``/etc/pylintrc``
+#. ``pylintrc``
+#. ``.pylintrc``
+#. ``pyproject.toml``, provided it has at least one ``tool.pylint.`` section.
+#. ``setup.cfg``, provided it has at least one ``pylint.`` section.
```
What happens if two or more valid files are found in the same location?
Maybe the doc should explain this.

danrneal replied:
I added that they were listed in order or priority.  So if .pylintrc and setup.cfg are in the same location it'll pick .pylintrc.  Is that clearer or do I need to expand more than just the note?

hippo91 commented on pylint/config/find_default_config_files.py:
```diff
@@ -24,40 +26,45 @@ def _cfg_has_config(path):
-
-    if os.path.isfile("/etc/pylintrc"):
-        yield "/etc/pylintrc"
+
```
Why did the `else` statement disappear?

danrneal replied:
Isn't the else is redundant when there is a yield or return statement inside and if block?

hippo91 reviewed (approved):
@danrneal thanks for this clear and welcomed PR!
I just have few comments/questions.

hippo91 commented on tests/test_config.py:
```diff
@@ -91,3 +92,16 @@ def test_can_read_toml_rich_types(tmp_path):
     check_configuration_file_reader(config_file)
+
+
+def test_gets_correct_config_files(tmp_path):
```
I cannot see where this test is called and how the `tmp_path` is set up. Are we sure this test is executed?

Pierre-Sassoulas replied:
`tmp_path` is a [pytest fixture](https://docs.pytest.org/en/stable/tmpdir.html#the-tmp-path-fixture)

hippo91 replied:
@Pierre-Sassoulas thanks for your answer. I should have seen that but i was a bit in a hurry.

Pierre-Sassoulas commented:
I'm moving this to 2.7.1, I think it's addable as a minor feature later.

AWhetter commented:
Sorry to jump in so late in the game. I don't think that it makes much sense to look for a pyproject.toml or a setup.cfg anywhere other than in the project because those files store metadata about a specific project. It's in the name of the pyproject file and in the PEP of the file (https://www.python.org/dev/peps/pep-0621/). Those files don't really belong outside of a project, so I can't think of a good reason to look for them anywhere other than the current directory.
Looking for a `.pylintrc` anywhere other than in the currently working directory also seems out of place. It makes sense to allow it in the current working directory because maybe you want to obscure it from view to keep the project tree looking tidy. But the directories `~/.config` and `/etc` are the places for config files, so I can't think of a reason why you would want to obscure a config file from view in those places.

Side note: For a while I've been feeling like the config system in pylint is not great and it needs a bit of attention. The code itself is pretty unwieldy and I think the user experience of configuration isn't great either. Changes like this make me nervous because some day we'll want to give the configuration the attention it deserves, but that means making breaking changes. So I'd rather not expand the bounds of the configuration system too much so that aren't breaking as much for people later on and so we can focus more on what the config system should do, rather than needing to worry about what we might break.

AWhetter reviewed (commented):
Sorry to jump in so late in the game. I don't think that it makes much sense to look for a pyproject.toml or a setup.cfg anywhere other than in the project because those files store metadata about a specific project. It's in the name of the pyproject file and in the PEP of the file (https://www.python.org/dev/peps/pep-0621/). Those files don't really belong outside of a project, so I can't think of a good reason to look for them anywhere other than the current directory.
Looking for a .pylintrc anywhere other than in the currently working directory also seems out of place. It makes sense to allow it in the current working directory because maybe you want to obscure it from view to keep the project tree looking tidy. But the directories ~/.config and /etc are the places for config files, so I can't think of a reason why you would want to obscure a config file from view in those places.

Side note: For a while I've been feeling like the config system in pylint is not great and it needs a bit of attention. The code itself is pretty unwieldy and I think the user experience of configuration isn't great either. Changes like this make me nervous because some day we'll want to give the configuration the attention it deserves, but that means making breaking changes. So I'd rather not expand the bounds of the configuration system too much so that aren't breaking as much for people later on and so we can focus more on what the config system should do, rather than needing to worry about what we might break.

Pierre-Sassoulas commented:
@AWhetter  I agree that the configuration need to be redesigned and that we don't want to have too much to break when we do that. I also agree pyproject.toml or setup.cfg makes no sense outside of the current project directory and you convinced me we should not start maintaining and documenting things that do not make sense (especially considering the fact that we had two possible names for pylintrc already ``.pylintrc`` and ``.pylintrc``).

But isn't `etc/` or `~/.config` the right place for a global pylint configuration used for all projects on a computer ? I was thinking of maybe using ``confuse`` for the configuration redesign and they put their configuration file in `XDG_CONFIG_HOME ` :

>  Look for configuration files in platform-specific paths. Like $XDG_CONFIG_HOME or ~/.config on Unix; “Application Support” on macOS; %APPDATA% on Windows. Your program gets its own directory, which you can use to store additional data. You can transparently create this directory on demand if, for example, you need to initialize the configuration file on first run. And an environment variable can be used to override the directory’s location.

Source: https://pypi.org/project/confuse/

AWhetter commented:
> But isn't etc/ or ~/.config the right place for a global pylint configuration used for all projects on a computer

Yes absolutely! I like both of these places for a pylintrc file. But not for `.pylintrc`, `pyproject.toml`, or `setup.cfg`.

Pierre-Sassoulas commented:
I think the `pylintrc` could also be global ? For example you have a default value in pylint, you can change it in `etc/` with a pylintrc file, then you can change both in your repository with the local file. I get it would means we need an option for a local config to signal that it's bypassing the global config for all value and define pylint's default unless explicitly overridden. Or do you want to change the name of the file during the refactor ?

AWhetter commented:
Yes that's what I meant. `pylintrc` should be global. `.pylintrc`, `pyproject.toml`, and `setup.cfg` should no be global.

I definitely think that config inheritance is something that we should consider in the future, but we shouldn't implement it now.

danrneal commented:
It seems like this change is no longer the direction to go.  My original example was isort searching for pyproject.toml in the home directory. I thought would be nice if there were a single global default file for black, pylint, isort.  However, isort and black have both removed this capability so pylint certainly should *not* add this.

I'm not sure there is any value in me addressing the comments.  I'll wait to hear back from project owners actions to take.  Feel free to either close this PR or if you would like me to amend it, let me know and I'll try to address existing comments and make any changes to it soon.

danrneal commented:
Both isort and black stop going up directories when they get to a .git or .hg directory.  But the use case for having a single config for isort, pylint, and black in the home directory is for coding that isn't part of a project.  If you had a global default pyproject.toml  then your code can still be linted and checked for style even if your code is not part of a project.

I can address the comments in the PR and get it back into ship shape.  However, if this isn't the direction to go in, then there probably isn't a lot of value in me doing that.  I'll wait for more discussion.

Pierre-Sassoulas commented:
I think we could mirror what isort and black do, what do you think @AWhetter ? Maybe we need to refactor this for 3.0 as a real breaking change ?

Pierre-Sassoulas commented:
I'm going to close as we're going to move to non global configuration files. Thank you for your contribution and sorry for being indecisive @danrneal
